### PR TITLE
fix(math): replay certified Smith transformation relations against their source

### LINKED ISSUE #2285: [Bug]: Certified Smith results do not replay their transformation relation
**Priority**

P1

**Area**

checker / exact mathematical result contract

**Component**

`matrix.normal_form.smith.certified.compute`

**Observed behavior**

`SmithNormalFormCertificate` describes an exact transformation certificate `D = U A V` with unimodular `U` and `V`, but its result validator checks only matrix shapes and the canonical divisibility diagonal. It does not replay `D = U A V` or verify that the declared determinant signs equal the actual determinants of `U` and `V`.

As a result, the following inconsistent authored certificate validates: `A = [2]`, `D = [2]`, `U = [2]`, `V = [1]`, with both declared determinant signs equal to `1`. In fact `U A V = [4]` and `det(U) = 2`.

The producer currently checks the relation and determinants before constructing its result. That does not make the public result contract fail closed when evidence is independently authored, deserialized, mutated, or consumed downstream.

**Expected behavior**

`SmithNormalFormCertificate` should validate all of its advertised exact relations:

- `diagonal == left_transformation * source * right_transformation`;
- the actual determinant of each square transformation is the declared `-1` or `1`;
- both transformations are therefore unimodular;
- rank and invariant factors agree with the reconstructed diagonal.

Mutation of any source, diagonal, transformation, determinant, rank, or invariant-factor field should fail unless the complete certificate remains valid.

**Steps to reproduce**

On current `origin/main` (`96fe5893f1468f62f9b08b27b0215740c58b50d6`), with Python 3.12.13:

```python
from jacobian.math.matrices.certified_snf.values import (
    CertifiedIntegerMatrix,
    SmithNormalFormCertificate,
)

def matrix(value: int) -> CertifiedIntegerMatrix:
    return CertifiedIntegerMatrix(
        row_count=1,
        column_count=1,
        entries=((str(value),),),
    )

certificate = SmithNormalFormCertificate(
    source=matrix(2),
    diagonal=matrix(2),
    left_transformation=matrix(2),
    right_transformation=matrix(1),
    rank=1,
    invariant_factors=("2",),
    left_determinant="1",
    right_determinant="1",
)

print(certificate.diagonal.entries)  # (("2",),)
# But U*A*V = 2*2*1 = 4, and det(U) = 2.
```

**Additional context**

#581 explicitly required full `U * A * V = D`, unimodularity, rank, and invariant-factor checks to remain green. The producer performs those checks, but the public semantic value does not.

Tests currently cover shape, canonical diagonal rejection, and source handoff. Add result-contract mutation tests for each matrix and determinant field, including rectangular, rank-deficient, and explicit zero-dimensional cases.

### PR DESCRIPTION
**Pull request by morluto**
## Problem

`SmithNormalFormCertificate` advertises the exact transformation certificate `D = U A V` with unimodular `U` and `V`, but its validator checked only matrix shapes and the canonical divisibility diagonal. It never replayed `D = U A V` nor verified that the declared determinant signs equal the actual determinants, so an independently authored, deserialized, mutated, or downstream-consumed certificate such as `A=[2]`, `D=[2]`, `U=[2]`, `V=[1]` with declared signs `1/1` validated — despite `U A V = [4]` and `det(U) = 2`. The producer's internal pre-construction checks did not make the public result contract fail closed (#2285).

## Root cause

The semantic value's validation stopped at structural coherence (shapes, diagonal canonicality, divisibility chain) and treated the declared relation/determinant fields as evidence instead of replaying the defining relations against the retained source and transformation matrices.

## Fix

- Add a `bind_relation_to_source` model validator to `SmithNormalFormCertificate` that replays `diagonal == left_transformation * source * right_transformation` exactly over the parsed canonical integers and recomputes each square transformation's exact determinant over `ZZ`, requiring it to equal the declared `-1`/`1` (which establishes unimodularity). Rank and invariant factors already had to match the retained diagonal, so they are now bound to the reconstructed relation too.
- Replay uses lazy imports of the owner's domain kernel (`matrix_multiply`, new exact `matrix_determinant` over SymPy `ZZ`) per the established certified-result pattern; parsing/formatting uses the limit-free canonical integer helpers so large canonical values (e.g. 5001-digit factors) keep validating.
- The large-canonical-invariant-factor boundary test now retains a self-consistent certificate (`source == diagonal`), since its previous form (`source=[0]`, `diagonal=[10^5000]`) is exactly the kind of inconsistent authored value that must now fail.

## Validation

- Regression tests in `tests/math/matrices/test_certified_snf_models.py`: the reported reproduction, fail-closed mutation of every matrix, determinant, rank, and invariant-factor field, determinant-sign mismatch against a genuinely unimodular `-1` transformation, rectangular, rank-deficient, and explicit zero-dimensional (`0×3`, `2×0`) boundaries.
- `uv run --locked pytest`: 4658 passed.
- `uv run ruff check` + `ruff format --check`: clean on changed files.
- `uv run mypy src/jacobian/math/matrices/certified_snf`: no issues.

Fixes #2285

### CODE REVIEW

File: src/jacobian/math/matrices/certified_snf/values.py [hunk 1]
```diff
@@ -116,5 +116,56 @@ def require_coherent_shapes_and_canonical_diagonal(self) -> Self:
...
+                "Smith certificate transformations must replay "
+                "diagonal = left * source * right exactly"
+            )
+        for label, transformation, determinant in (
+            ("left", left, self.left_determinant),
+            ("right", right, self.right_determinant),
+        ):
+            actual_determinant = format_canonical_integer(
+                matrix_determinant(transformation)
+            )
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/matrices/certified_snf/values.py]: Root cause: the replay validator formatted each transformation determinant to its full decimal expansion with format_canonical_integer solely to string-compare against the declared "1"/"-1" literal, then embedded that expansion in the rejection error. For a rank-zero certificate with a 32x0 source the relation check is trivially true, so any 32x32 transformation with entries up to 32768 digits reached this code and forced a ~million-digit repeated-division conversion plus an unbounded error payload before rejection.

Fix: values.py now compares matrix_determinant(transformation) numerically against int(determinant), so no decimal formatting of the computed determinant ever happens; the diagnostic keeps only the bounded declared sign ("must be the declared unimodular -1") and drops the actual value. The unused format_canonical_integer import was removed.

Validation: tests/math/matrices/test_certified_snf_models.py::test_certificate_contract_replays_rank_zero_determinants_without_formatting_them (known-answer acceptance of a det -1 rank-zero certificate plus adversarial rejection whose message provably omits the 64-digit determinant); full tests/math/matrices (145 passed), ruff and mypy clean on the touched files.
